docs/proposals/20210222-kubelet-authentication.md

```diff
@@ -0,0 +1,343 @@
+
+#### Kubeadm
+Since this proposal essentially takes over part of the node registration process from kubeadm, we will require the following changes:
+- kubeadm COULD allow opt-out of kubeadm setting up ClusterRoleBindings between the system:nodes group and the `system:certificates.k8s.io:certificatesigningrequests:selfnodeclient` permission, so that certificate renewals must go through re-attestation.
+- Kubeadm COULD allow opt-out of kubeadm setting up `kubeadm:node-autoapprove-bootstrap` cluster role binding.
+This is deferred to a future Kubeadm design and release, and for this proposal, we will add fields to KubeadmControlPlane to remove these node groups and bindings post control plane initialisation.
+
+The idea is to rely the [exec mechanism](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#configuration) of kubeconfigs with local cache directory, when kubelet wants to talk to the apiserver it will call on `nodeadm` to get a client certificate.
+
+#### Changes to KubeadmControlPlane
+
+```
+spec:
+  disableKubeadmAutoApproveClusterRoleBinding: true
+```
+
+This field on KCP will `kubeadm:node-autoapprove-bootstrap` cluster role binding.
+
+#### Changes to CAPBK
+
+the bootstrap process will change to accomodate the new attestation process and ensure backward compatibility, in particular, the bootstrapping will look if `nodeadm` is present on the machine. If it's the case, the bootstrap would use node attestation. If `nodeadm` is not available, we would fallback on older bootstrapping flow. This would enable supporting the existing/older kubernetes versions and machine images built for it.
+
+#### Kubelet authenticator flow
+
+The authenticator will be responsible for both updating the kubelet client and serving certificates.
+
+Both of these decision flows will operate inside the authenticator, and execute whenever called by kubelet to provide client authentication.
+
+##### Client authentication flow
+
+![client auth](images/kubelet-authentication/client-authenticator-flow.png)
+
+##### Serving certificate flow
+
+![serving cert flow](images/kubelet-authentication/serving-certificate-flow.png)
+
+### Risks and Mitigations
+
+There may be additional security risks being introduced in this design. In order to mitigate this, this proposal will be taken to SIG Security and SIG Auth for review **before the beta graduation**.
+
+## Alternatives
+
+#### Implement within the cloud providers instead of Cluster API
+Given that there is an existent implementation in cloud-provider-gcp, this could be extended to all of the cloud providers. However, there are some advantages to making Cluster API responsible for kubelet registration in that no changes to the assumptions around connectivity between management and workload clusters are required, neither does the signer need to be included as a static pod during control plane instantiation.
```

Comment:
Implementing within cluster API also allows mixing cloud and on-premises instances within a single mechanism, which could be useful. It goes without saying that bare-metal on premises systems can't rely on the cloud provider.